### PR TITLE
initial commit of skywalking-rust with full-tracing features.

### PR DESCRIPTION
**Pull request by Shikugawa**

### CODE REVIEW

**Comment by Shikugawa** [comment]
@wu-sheng PTAL

**Comment by wu-sheng** [comment]
Whoo, you have finished this.

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -10,20 +10,15 @@ on:
 
 
 jobs:
-  CI:
-    runs-on: ubuntu-18.04
-    timeout-minutes: 180
-    strategy:
-      fail-fast: true
+  test:
```
Comment (wu-sheng) [inline .github/workflows/ci.yaml:13]: I think this rename could make use unable to merge. ASF Infra group sets up `CI` as required task in default.
We could change the settings through [this file](https://github.com/apache/skywalking-rust/blob/master/.asf.yaml).

**Comment by wu-sheng** [comment]
BTW, we should add our own `Attributes` to support adding a span in the current context, then use attribute to add tag or use explicit APIs to add tags/logs

WDYT?

File: README.md [hunk 2]
```diff
@@ -30,60 +30,24 @@ Tag and Log are similar attributes of the span.
...
+async fn main() {
+    let tx = Reporter::start("http://0.0.0.0:11800".to_string()).await;
+    let mut context = TracingContext::default("service", "instance");
     {
-        assert_eq!(span2.span_id(), 1);
-        let mut span3 = context.create_exit_span("op3", Some(&span2), "127.0.0.1:8080", Some(&dyn extractor));
-        assert_eq!(span3.span_id(), 2);
-
-        context.finish_span(span3);
+        let span = context.create_entry_span("op1").unwrap();
```
Comment (wu-sheng) [inline README.md:45]: Should we have more API doc about spans, tags and logs?

**Comment by Shikugawa** [comment]
> BTW, we should add our own `Attributes` to support adding a span in the current context, then use attribute to add tag or use explicit APIs to add tags/logs

How is `Attributes`? I expect it works to register default tags/logs settings to span, is it correct?

**Comment by wu-sheng** [comment]
We could start with adding a local tag? And tag/log and other types of span.

**Comment by wu-sheng** [comment]
This is a suggestion, we could do that in another PR later.
Just fixing the CI task name and doc(APIs) should be good.

**Comment by Shikugawa** [comment]
> We could start with adding a local tag? And tag/log and other types of span.

Ah ok. What you said is like [this](https://github.com/SkyAPM/go2sky/blob/master/span.go#L148-L157)?

**Comment by wu-sheng** [comment]
Oh, I mean this https://doc.rust-lang.org/reference/procedural-macros.html. Something like

```rust
#[span(xxx)]
pub fn one_function(....) -> .... {
    ....
}
```

**Comment by Shikugawa** [comment]
> Oh, I mean this https://doc.rust-lang.org/reference/procedural-macros.html. Something like
> 
> ```rust
> #[span(xxx)]
> pub fn one_function(....) -> .... {
>     ....
> }
> ```

Got it. SG and worth doing to add this feature.

File: README.md (hunk 2 above)
Comment (Shikugawa) [reply README.md:45]: added

File: src/context/trace_context.rs [hunk 3]
```diff
@@ -25,11 +25,42 @@ use std::sync::Arc;
 
 use super::system_time::UnixTimeStampFetcher;
 
+/// Span is a concept that represents trace information for a single RPC.
+/// The Rust SDK supports Entry Span to represent inbound to a service
+/// and Exit Span to represent outbound from a service.
+///
+/// # Example
```
Comment (wu-sheng) [inline src/context/trace_context.rs:32]: This should have a copy on the readme.md in the root, right?

Comment (Shikugawa) [reply src/context/trace_context.rs:32]: Yes. fixed.

**Comment by wu-sheng** [comment]
CI fails.

File: tests/e2e/data/_expected_context.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
+segmentItems:
```
Comment (wu-sheng) [inline tests/e2e/data/_expected_context.yaml:1]: `tests/e2e/data/_expected_context.yaml` This file seems a temp file?

**Comment by wu-sheng** [comment]
You seems forgot to add e2e into GHA control file.

**Comment by Shikugawa** [comment]
> You seems forgot to add e2e into GHA control file.

I'm debugging e2e CI in my local repository. I will send another PR if it is ready.

File: tests/e2e/data/_expected_context.yaml (hunk 4 above)
Comment (Shikugawa) [reply tests/e2e/data/_expected_context.yaml]: Yes. deleted.

**Review by wu-sheng — APPROVED** [review]

**Pull request merged**